src/wp-includes/media.php

```diff
@@ -5672,9 +5672,10 @@ function wp_get_loading_optimization_attributes( $tag_name, $attr, $context ) {
 
 	/*
 	 * The first elements in 'the_content' or 'the_post_thumbnail' should not be lazy-loaded,
-	 * as they are likely above the fold.
+	 * as they are likely above the fold. Shortcodes are processed after content images, so if
+	 * thresholds haven't already been met, apply the same logic to those as well.
 	 */
-	if ( 'the_content' === $context || 'the_post_thumbnail' === $context ) {
+	if ( 'the_content' === $context || 'the_post_thumbnail' === $context || 'do_shortcode' === $context ) {
```

Comment:
We should add test coverage for this. I think most importantly it would be good to have an integration test, e.g. covering that a post that contains some images and a gallery shortcode has the images counted in the expected order and as such lazy-loaded / receiving fetchpriority as expected. If we want to be thorough we could also add a unit test directly calling `wp_get_loading_optimization_attributes()` with the `do_shortcode` context to make sure that that yields the expected return value.

Reply:
+1 for unit test.

Reply:
Unit tests added in [87d12ca](https://github.com/WordPress/wordpress-develop/pull/4799/commits/87d12cadd5f2e6f01ba6a1d328e822eff33cbaa4). I'm not sure what test case would include nested shortcodes but will try to think of one that we can add.